docs/src/writing_good_rules.md

```diff
@@ -103,16 +103,197 @@ In fact, any number of scalar arguments is supported, as is returning a tuple of
 See docstrings for the comprehensive usage instructions.
 ## Write tests
 
-In [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl)
-there are fairly decent tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
-Take a look at existing [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl) tests and you should see how to do stuff.
+[ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl)
+provides tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
+Take a look at the documentation or the existing [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl) tests to see how to write the tests.
+It's easy!
 
 !!! warning
-    Use finite differencing to test derivatives.
     Don't use analytical derivations for derivatives in the tests.
     Those are what you use to define the rules, and so can not be confidently used in the test.
     If you misread/misunderstood them, then your tests/implementation will have the same mistake.
+    Use finite differencing methods instead, as they are based on the primal computation.
 
 ## CAS systems are your friends.
 
 It is very easy to check gradients or derivatives with a computer algebra system (CAS) like [WolframAlpha](https://www.wolframalpha.com/input/?i=gradient+atan2%28x%2Cy%29).
+
+## Which functions need rules?
+
+In principle, a perfect AD system only needs rules for basic operations and can infer the
+rules for more complicated functions automatically.
```

Comment:
minor style comment: in markdown docs, writing one sentence per line (rather than having a character limit for lines) makes it easier to both make suggestions and to review diffs in future PRs